probabl-ai/skore
# feat: Implement basic `cross_validate` functionality

Addresses the first part of #383 

Introduces a new item type, `CrossValidateItem`, as well as a top-level function `skore.cross_validate`, which detects the type of ML task based on the input estimator and target, adds some scorers, and runs scikit-learn's `cross_validate` with those extra arguments.
The results are stored in the `Project` given as input, always under key `cross_validation`, along with an interactive summary plot.
The output of `skore.cross_validate` is as close as possible to scikit-learn's, except for some edge cases. Also, if the user passes `scorer="my_metric"`, the output dict will contain key `test_my_metric` as well as `test_score`, whereas in the scikit-learn version the dict only has `test_score`.

TL;DR:

- Make Plotly a mandatory dependency
- Add `CrossValidateItem`
- Add `cross_validate` function
- Add `plot_cross_validation` function that plots the result of a cross-validation in Plotly
- Refactor tests (add fixture `in_memory_project` and use that instead of `project` for clarity -- it would have been great to do `from conftest import in_memory_project as project` but pytest doesn't work that way)
- Fix some issues with VegaWidget in skore-ui (the plot was originally written in altair)

---

How to test:
  - I'm in a notebook/VSCode, I have a sklearn estimator, I run `skore.cross_validate` on it just like I would with `sklearn.model_selection.cross_validate`, except I also give the argument `project=project` to let Skore know where to store things (if the argument is not passed we'll do cross-validation but nothing will be saved)
  - A plot appears in the UI and in my notebook
  - I forgot a metric; I change the cell and re-run
  - A new plot is outputted and the one on the UI updates

---

To-do:
- [x] Add Marie's tests as automated tests
- [ ] Potentially use `pytest-cases` for cross-validation tests
- [x] Deal with multi-class classification (use `["roc_auc_over_weighted", "neg_log_loss", "recall_weighted", "precision_weighted"]`)
- [x] Deal with the case where `X` and `y` are not numpy arrays
  - At least set up the infrastructure to easily support pandas, numpy and lists; if only numpy works for now, that's okay
  - Look into `struct.pack`
- [x] Remove metrics depending on e.g. if estimator has `predict_proba`
  - ~Note for example SVC has the method defined, but it raises if the model isn't initialized with `probability=True`, so a `try-except` will be more accurate than `hasattr`~
- [x] Deal with the case where the estimator params are not serializable
  - Use `repr` for the time being
- [x] Make it so when the user passes `metric="my_metric"`, the output contains `test_my_metric` as well as `test_score`
- [x] Ensure that the stored cv_results are serializable
  - [x] Serialize the numpy arrays
  - [x] Deal with the case where the user passes `return_estimator=True` (we can't pass that to `CrossValidateItem` because it's not serializable)
    - Asked Marie
  - [ ] Confirm that the `get()` output is appropriate
- [x] Fix plot not updating in the UI
- [x] Remove default metrics for clustering
- [x] Stop storing the plot directly in `CrossValidateItem`, instead store JSON
  - Rename the attribute to `plot_serialized`, add a property `plot`
- [ ] ~Investigate Marie's issue of the empty plot~
